components/blitz/src/omero/gateway/facility/ROIFacility.java

```diff
@@ -834,71 +846,190 @@ else if (shape.isDirty()) {
      * @throws DSOutOfServiceException
-     *             If the connection is broken, or logged in.
      * @throws DSAccessException
-     *             If an error occurred while trying to retrieve data from OMEDS
-     *             service.
      */
-    public Collection<ROIResult> loadROIsForFolder(SecurityContext ctx,
-            long imageId, long folderId) throws DSOutOfServiceException,
+    public Collection<ROIData> loadROIs(SecurityContext ctx, Long imageId,
+            Collection<Long> folderIds) throws DSOutOfServiceException,
             DSAccessException {
         try {
-            // TODO: This should actually happen on the server; replace
-            //      with server-side method when available
-            
-            // get all ROIResults
-            List<ROIResult> roiresults = loadROIs(ctx, imageId);
+            List<Long> roiIDs = new ArrayList<Long>();
+
+            if (folderIds == null) {
+                // ignore folders, ROIs can but don't have to be part of folders
+                String query;
+                ParametersI p = new ParametersI();
+                if (imageId == null) {
+                    // ignore imageId, get all rois
+                    query = "select distinct roi from Roi roi";
+                } else if (imageId < 0) {
+                    // get all rois not attached to any image
+                    query = "select distinct roi from Roi roi "
+                            + "where roi.image is null";
+                } else {
+                    // get all rois for the image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.image as image "
+                            + "where image.id = :id";
+                    p.addId(imageId);
+                }
 
-            // get the ROIs of the specified folder
-            IQueryPrx qs = gateway.getQueryService(ctx);
-            StringBuilder sb = new StringBuilder();
-            ParametersI param = new ParametersI();
-            param.addLong("folderId", folderId);
+                IQueryPrx service = gateway.getQueryService(ctx);
+
+                List<IObject> objs = service.findAllByQuery(query, p);
+
+                for (IObject obj : objs)
+                    roiIDs.add(((Roi) obj).getId().getValue());
+            } else if (folderIds.isEmpty()) {
+                // ROIs must not be part of folders
+
+                String query;
+                ParametersI p = new ParametersI();
+                if (imageId == null) {
+                    // ignore imageId, get all rois
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.folderLinks as links ";
+                } else if (imageId > -1) {
+                    // get rois for the image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.image as image "
```

Comment:
interesting that you do the join fetch image here but not for the query above -- if you need it just for the where, you can simply write "where roi.image.id = :id"